api/src/wfl/service/postgres.clj

```diff
@@ -67,13 +57,22 @@
            (remove (comp finished? :status))
            (run! #(update! % (get-status! workload %)))))))
 
-(def update-workflow-statuses!
-  "Use `tx` to update `status` of Cromwell `workflows` in a `workload`."
-  (letfn [(get-cromwell-status [{:keys [cromwell]} {:keys [uuid]}]
-            (if (util/uuid-nil? uuid)
-              "skipped"
-              (cromwell-status cromwell uuid)))]
-    (make-update-workflows get-cromwell-status)))
+(defn update-workflow-statuses!
+  "Use `tx` to update `status` of unfinished Cromwell `workflows` in a `workload`."
+  [tx {:keys [cromwell workflows] :as workload}]
+  (let [ws-to-update (->> workflows
+                          (remove (comp nil? :uuid))
+                          (remove (comp finished? :status)))
+        uuid->status (if (empty? ws-to-update)
+                       (constantly nil)
+                       (->> ws-to-update
+                            (map (fn [{:keys [uuid]}] {:id uuid}))
+                            (#(conj % {:includeSubworkflows "false"}))
+                            (cromwell/query (first (all/cromwell-environments cromwell)))
+                            (map (fn [{:keys [id status]}] [id status]))
+                            (into {})))
+        get-status   (fn [_ {:keys [uuid]}] (or (uuid->status uuid) "skipped"))]
+    ((make-update-workflows get-status) tx workload)))
```

Comment:
For the first point, sure, I'll change it around to query for the label. I had done `id` because I assumed it would be more performant and it works real nice with AoU stuff for not requesting stuff that is already finished.

For the second point, I'm not sure how `make-update-workflows` doesn't fit that, it just maps over the workflows. What do you mean by grouping by uuid?

Reply:
Either 
- pair up the workflows stored in the database with those returned by querying cromwell by workload UUID (can you guarantee they're in the same order)
- update the workflow, adjusting sql update where clause to use the workflow uuid

`make-update-workflows` does neither of these things.

Reply:
Oh, I thought we'd want to query the DB by primary key--just not used to DB stuff changing it now

Reply:
We do want to use primary keys where possible. The issue is I'm not sure you can guarantee that we've stored workflows in the same order as cromwell returns them in the "get by label" query.

Reply:
You are correct. The code always assumed the order was random. This was the purpose of using `make-update-workflows`, it allows us to iterate over the workflows, accessing both their UUID field and ID key at the same time. The `get-status!` function passed to `make-update-workflows` simply converted the UUID for that row into the status. The `get-status!` function was literally just a map of UUIDs to statuses, at each row we wanted to update we'd query the map and fill in a status if there was one.

Reply:
No it didn't - it looped through each workflow, called cromwell about that workflow, then updated the workflow by its primary key in the database.